### PR TITLE
Refactor of shader program related stuff

### PR DESCRIPTION
**Pull request by Mrkol**
- Use best post-modern C++ practices: view types in favour of const references (string_view, span, etc). Added a new StringMap alias for string_view-based lookup to work with unordered_map.
- For paths, replaced string with filesystem::path. It should theoretically support cyrillic characters.
- Removed trailing spaces in a bunch of places.
- Replaced old `if (xxx) ETNA_PANIC(...)` stuff with ETNA_ASSERT. Also replaced cout logging with spdlog logs.
- Flatten dependencies of ShaderProgramManager. Accessing stuff from GlobalContext within etna is a bad idea IMO, as that header includes basically all of etna, so compile times will get pretty bad pretty quickly. This approach also makes initialization and cleanup order less brittle, as the constructor explicitly takes all dependency references.

### CODE REVIEW

**Comment by AlexandrShcherbakov** [comment]
So many changes in one commit. It is difficult to review.

File: etna/include/etna/ShaderProgram.hpp [hunk 1]
```diff
@@ -24,19 +25,18 @@ namespace etna
...
-    const std::string &getName() const { return entryPoint; }
+    std::string_view getName() const { return entryPoint; }
+    const char* getNameCstr() const { return entryPoint.c_str(); }
     vk::PushConstantRange getPushConst() const { return pushConst; }
 
-    ShaderModule(const ShaderModule &mod) = delete;
-    ShaderModule &operator=(const ShaderModule &mod) = delete;  
   private:
     std::string path {};
-    std::string entryPoint {};
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/ShaderProgram.hpp:34]: А зачем?

File: etna/include/etna/ShaderProgram.hpp [hunk 2]
```diff
@@ -5,6 +5,7 @@
 #include <etna/Vulkan.hpp>
 #include <etna/Forward.hpp>
 #include <etna/DescriptorSetLayout.hpp>
+#include <etna/detail/StringMap.hpp>
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/ShaderProgram.hpp:8]: Что здесь значит detail?

File: etna/include/etna/ShaderProgram.hpp [hunk 3]
```diff
@@ -102,10 +102,10 @@ namespace etna
     ShaderProgramManager &operator=(const ShaderProgramManager &) = delete;
 
   private:
-    std::unordered_map<std::string, uint32_t> shaderModuleNames;
+    std::unordered_map<std::filesystem::path, uint32_t> shaderModuleNames;
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/ShaderProgram.hpp:114]: shaderModulePaths

File: etna/include/etna/DescriptorSet.hpp [hunk 4]
```diff
@@ -16,7 +16,7 @@ namespace etna
       : generation {gen}, layoutId {id}, set {vk_set} {}
 
     bool isValid() const;
-
```
Comment (AlexandrShcherbakov) [inline etna/include/etna/DescriptorSet.hpp:19]: Лучше бы этот коммит был первым.

File: etna/include/etna/GlobalContext.hpp [hunk 5]
```diff
@@ -53,8 +53,8 @@ namespace etna
 
     std::unique_ptr<VmaAllocator_T, void(*)(VmaAllocator)> vmaAllocator{nullptr, nullptr};
 
-    DescriptorSetLayoutCache descriptorSetLayouts {}; 
-    ShaderProgramManager shaderPrograms {};
+    DescriptorSetLayoutCache descriptorSetLayouts {};
+    std::optional<ShaderProgramManager> shaderPrograms;
```
[the remark this answers is not in the document]
Comment (Mrkol) [reply etna/include/etna/GlobalContext.hpp:62]: For late initialization. All of these optional fields depend on vulkan initialization, so it's easier to make them optional and init at the end of the constructor

File: etna/include/etna/ShaderProgram.hpp (hunk 2 above)
Comment (AlexandrShcherbakov) [reply etna/include/etna/ShaderProgram.hpp:8]: Лучше internal.

**Pull request closed without merging**